packages/components/tooltip/src/tooltip.vue

```diff
@@ -1,3 +1,177 @@
+  ...createModelToggleComposableResult.useModelToggleEmits,
+  'before-show',
+  'before-hide',
+  'show',
+  'hide',
+  'open',
+  'close',
+])
+
+const { useModelToggle } = createModelToggleComposableResult
+
+usePopperContainer()
+const compatShowAfter = computed(() => {
+  if (!isUndefined(props.openDelay)) {
+    debugWarn(
+      'ElTooltip',
+      'open-delay is about to be deprecated in the next major version, please use `show-after` instead'
+    )
+  }
+  return props.openDelay || (props.showAfter as number)
+})
+const compatShowArrow = computed(() => {
+  if (!isUndefined(props.visibleArrow)) {
+    debugWarn(
+      'ElTooltip',
+      '`visible-arrow` is about to be deprecated in the next major version, please use `show-arrow` instead'
+    )
+  }
+  return isBoolean(props.visibleArrow) ? props.visibleArrow : props.showArrow
+})
+
+const id = useId()
+// TODO any is temporary, replace with `InstanceType<typeof ElPopper> | null` later
+const popperRef = ref<any>(null)
+// TODO any is temporary, replace with `InstanceType<typeof ElTooltipContent> | null` later
+const contentRef = ref<any>(null)
+
+const updatePopper = () => {
+  const popperComponent = unref(popperRef)
+  if (popperComponent) {
+    popperComponent.popperInstanceRef?.update()
+  }
+}
+const open = ref(false)
+const toggleReason = ref<Event | undefined>(undefined)
+
+const { show, hide, hasUpdateHandler } = useModelToggle({
+  indicator: open,
+  toggleReason,
+})
+
+const { onOpen, onClose } = useDelayedToggle({
+  showAfter: compatShowAfter,
+  hideAfter: toRef(props, 'hideAfter'),
+  open: show,
+  close: hide,
+})
+
+const controlled = computed(
+  () => isBoolean(props.visible) && !hasUpdateHandler.value
+)
+
+provide(TOOLTIP_INJECTION_KEY, {
+  controlled,
+  id,
+  open: readonly(open),
+  trigger: toRef(props, 'trigger'),
+  onOpen: (event?: Event) => {
+    onOpen(event)
+  },
+  onClose: (event?: Event) => {
+    onClose(event)
+  },
+  onToggle: (event?: Event) => {
+    if (unref(open)) {
+      onClose(event)
+    } else {
+      onOpen(event)
+    }
+  },
+  onShow: () => {
+    emit('show', toggleReason.value)
+  },
+  onHide: () => {
+    emit('hide', toggleReason.value)
+  },
+  onBeforeShow: () => {
+    emit('before-show', toggleReason.value)
+  },
+  onBeforeHide: () => {
+    emit('before-hide', toggleReason.value)
+  },
+  updatePopper,
+})
+
+watch(
+  () => props.disabled,
+  (disabled) => {
+    if (disabled && open.value) {
+      open.value = false
+    }
+  }
+)
+
+onDeactivated(() => open.value && hide())
+
+const isFocusInsideContent = () => {
+  const popperContent: HTMLElement | undefined =
+    contentRef.value?.contentRef?.popperContentRef
+  return popperContent && popperContent.contains(document.activeElement)
+}
+
+defineExpose({
+  compatShowAfter,
+  compatShowArrow,
+  popperRef,
+  contentRef,
+  open,
+  hide,
+  isFocusInsideContent,
+  updatePopper,
+  onOpen,
+  onClose,
+  disabled: props.disabled,
+})
```

Comment:
Is it necessary to expose all variables? We can remove them I think except `popperRef` and `contentRef`.